### PR TITLE
Lock Pattern #71: Added mutex and semaphore modules to demonstrate locks 

### PR DESCRIPTION
**Pull request by gwildor28**
Added two modules to demonstrate locks.

Mutex demonstrates a simple mutual exclusion lock. This is a binary lock which only allows one thread at a time to access the resource. See: [Locks on Wikipedia](https://en.wikipedia.org/wiki/Lock_(computer_science))

Semaphore demonstrates a semaphore for controlling access to a pool of resources. See [Semaphores on Wikipedia](https://en.wikipedia.org/wiki/Semaphore_(programming))

The main class of both programs is App.java.

I haven't created any documentation yet. Let me know if you want any changes or need some explanations of the code, etc.

### CODE REVIEW

**Comment by gwildor28** [comment]
Hi, just thought I'd check if anyone has had the opportunity to review this yet. i understand I haven't put any documentation in here yet but I just wanted to see if this is what you wanted to demonstrate the pattern

**Comment by npathai** [comment]
@gwildor28 It will be picked up for review soon.

**Comment by gwildor28** [comment]
OK thanks

**Comment by iluwatar** [comment]
Review comments:
- Mutex/App.java the pattern should be described here as well as the example implementation
- Mutex example contains no tests
- Mutex javadocs should be improved. They are not there just to satisfy the compiler. Please describe what each class and method does.

File: mutex/src/main/java/com/iluwatar/mutex/Jar.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+ * THE SOFTWARE.
+ */
+package com.iluwatar.mutex;
+
+/**
+ * Jar.
+ */
+public class Jar {
+
+  private Lock lock;
```
Comment (iluwatar) [inline mutex/src/main/java/com/iluwatar/mutex/Jar.java]: private final Lock lock;

File: mutex/src/main/java/com/iluwatar/mutex/Mutex.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package com.iluwatar.mutex;
+
+/**
+ * Mutex.
+ */
+public class Mutex implements Lock {
+  private Object owner = null;
```
Comment (iluwatar) [inline mutex/src/main/java/com/iluwatar/mutex/Mutex.java]: It is null by default

File: mutex/src/main/java/com/iluwatar/mutex/Thief.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+package com.iluwatar.mutex;
+
+/**
+ * Thief.
+ */
+public class Thief extends Thread {
+
+  private String name;
+
+  private Jar jar;
```
Comment (iluwatar) [inline mutex/src/main/java/com/iluwatar/mutex/Thief.java]: name and jar should be final

File: semaphore/src/main/java/com/iluwatar/semaphore/Customer.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+package com.iluwatar.semaphore;
+
+/**
+ * Customer.
+ */
+public class Customer extends Thread {
+    
+  private String name;
+  private FruitShop fruitShop;
+  private FruitBowl fruitBowl;
```
Comment (iluwatar) [inline semaphore/src/main/java/com/iluwatar/semaphore/Customer.java]: the private member are only set in the constructor so they should be final

**Comment by iluwatar** [comment]
@gwildor28 you have my review comments. Please comment on this thread once you've addressed them so we'll do another review.

**Comment by gwildor28** [comment]
I've addressed the items listed in the review comments in my commit 2 days ago

**Comment by iluwatar** [comment]
@gwildor28 there are merge conflicts. Can you pull the latest changes into your fork from the master?

**Comment by gwildor28** [comment]
I've merged your code into my branch, unfortunately Maven is complaining of problems in project dao. The first issue is a missing version tag in the pom.xml file, for junit-hierarchicalcontextrunner. Even after putting the tag in the file, Maven fails the tests for this project:

com.iluwatar.dao.AppTest - fails with a java.io.IOException related to the location of the database
com.iluwatar.dao.InMemoryCustomerDaoTest and com.iluwatar.dao.DbCustomerDaoTest - fail with java.lang.NoClassDefFoundError related to org/junit/internal/runners/rules/RuleFieldValidator
and de.bechte.junit.runners.validation.RuleValidator

Have you added any requirements for other software to be installed on this system?
The error in the pom.xml is the same problem that is failing the checks for this pull request

**Comment by npathai** [comment]
@gwildor28 I am the culprit for these errors. I will solve them first thing tomorrow. Sorry for the delay

**Comment by gwildor28** [comment]
No problem.

**Comment by gwildor28** [comment]
hi @npathai  have you been able to look at this yet?

**Comment by npathai** [comment]
@gwildor28 Not yet. Give me a day, have been a bit busy.

**Comment by gwildor28** [comment]
@npathai OK no problem

**Comment by gwildor28** [comment]
Hi @npathai just thought to say I'm actually doing this lock pattern as part of a third year undergraduate degree module (open-source software) where I have to contribute to an existing open source project, The report is due on the 24th April and I'll get substantially more marks if I'm able to show a pull request of mine that got accepted. Do you think you'll realistically be able to sort out the errors within the next few days? Thanks

**Comment by gwildor28** [comment]
All seems to have passed the checks now, please review. Thanks

**Comment by iluwatar** [comment]
Looking quite good now @gwildor28 but still the tests are insufficient. Can you add unit test for `Mutex.java` and `Semaphore.java`?

**Comment by gwildor28** [comment]
Hi @iluwatar I've done the tests, please review. Thanks

**Comment by iluwatar** [comment]
Good work @gwildor28 Great to see you pull this through! :smile:

**Comment by gwildor28** [comment]
Thanks

**Comment by iluwatar** [comment]
@all-contributors please add @gwildor28 for code

**Pull request merged**